osv/packagist_version.py

```diff
@@ -0,0 +1,154 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Ordered packagist version"""
+import re
+
+
+class PackagistVersion:
+  """
+  Follows the packagist version ordering, which is recommended to be semver, but
+  not enforced to be semver. The php standard version comparison code written in
+  C is located here: https://github.com/php/php-src/blob/master/ext/standard/versioning.c
+
+  The function first replaces _, - and + with a dot . in the version strings and
+  also inserts dots . before and after any non number so that for example
+  '4.3.2RC1' becomes '4.3.2.RC.1'. Then it compares the parts starting from
+  left to right.
+
+  If a part contains special version strings these are handled in the following order:
+  any string not found in this list < dev < alpha = a < beta = b < RC = rc < # < pl = p.
+  This way not only versions with different levels like '4.1' and '4.1.2' can be
+  compared but also any PHP specific version containing development state.
```

Comment:
Any other ambiguous/undocumented things to call out in the implementation here?

Reply:
There was one other issue with how it dealt with `#`, but I believe I have now fixed that so that it should have the same behavior